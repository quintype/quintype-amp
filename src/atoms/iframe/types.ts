import { Common } from "../common-types";

export interface IframeTypes extends Common {
  src?: string;
  sandbox?: string;
  children?: JSX.Element | JSX.Element[] | React.ReactChildren;
  srcdoc?: string;
  frameborder?: string;
  allowfullscreen?: string;
  allowpaymentrequest?: string;
  allowtransparency?: string;
  referrerpolicy?: string;
  inlineStyles?: object;
}
