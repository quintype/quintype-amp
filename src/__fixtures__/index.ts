import { config, publisherConfig, ampConfig, configOpts } from "./config.fixture";
import { textStory } from "./text-story.fixture";
import { storyWithManyJsEmbeds } from "./story-with-jsembeds.fixture";
import { relatedStories } from "./related-stories";
import { seo } from "./seo.fixture";
import { allElementsStory } from "./all-element-story.fixture";
import { liveBlogStory } from "./live-blog-story.fixture";

export {
  allElementsStory,
  config,
  publisherConfig,
  ampConfig,
  textStory,
  storyWithManyJsEmbeds,
  relatedStories,
  configOpts,
  seo,
  liveBlogStory
};
