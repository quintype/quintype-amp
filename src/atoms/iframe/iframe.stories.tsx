import * as React from "react";
import { storiesOf } from "@storybook/react";
import { Iframe } from "./iframe";

const postSrc = "https://codepen.io/rvgpl/embed/qXGqKm?height=265&theme-id=dark&default-tab=css,result";

storiesOf("Iframe", module)
  .addDecorator((story) => (
    <div>
      <div style={{ height: "600px", display: "grid", placeItems: "center" }}>
        Amp iframe can't render 600px of the viewport, please scroll.
      </div>
      {story()}
    </div>
  ))
  .add("Post with 16:9 responsive (default) layout", () => <Iframe src={postSrc} />)
  .add("Post with fixed (custom) layout", () => <Iframe width="400" height="400" layout="fixed" src={postSrc} />)
  .add("Post with custom style", () => <Iframe src={postSrc} inlineStyles={{ border: "5px solid red" }} />);
