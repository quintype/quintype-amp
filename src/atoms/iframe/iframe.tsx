import React, { Fragment } from "react";
import { IframeTypes } from "./types";
import Helmet from "react-helmet";
import styled from "styled-components";

const StyledIframe = styled("amp-iframe").attrs(({ inlineStyles }: StyledIframeTypes) => ({
  style: inlineStyles
}))<StyledIframeTypes>``;

export interface StyledIframeTypes {
  inlineStyles?: object;
}

const Iframe = ({
  src,
  sandbox = "allow-scripts allow-same-origin",
  layout = "responsive",
  frameborder = "0",
  width = "16",
  height = "9",
  children,
  inlineStyles,
  ...restProps
}: IframeTypes) => {
  return (
    <Fragment>
      <Helmet>
        <script async={undefined} custom-element="amp-iframe" src="https://cdn.ampproject.org/v0/amp-iframe-0.1.js" />
      </Helmet>
      <StyledIframe
        inlineStyles={inlineStyles}
        width={width}
        height={height}
        sandbox={sandbox}
        layout={layout}
        frameborder={frameborder}
        src={src}
        {...restProps}>
        {children && children}
      </StyledIframe>
    </Fragment>
  );
};

export { Iframe };
