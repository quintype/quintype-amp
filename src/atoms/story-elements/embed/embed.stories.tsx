import * as React from "react";
import { storiesOf } from "@storybook/react";
import { Embed } from "./embed";
import { config, textStory } from "../../../__fixtures__";
import { Layout } from "../../layout";

const sampleJsEmbedElement = {
  description: "",
  "embed-js":
    "PGlmcmFtZSBoZWlnaHQ9IjI2NSIgc3R5bGU9IndpZHRoOiAxMDAlOyIgc2Nyb2xsaW5nPSJubyIgdGl0bGU9ImxpZ2h0IGJ1bGIiIHNyYz0iaHR0cHM6Ly9jb2RlcGVuLmlvL3ZsYWRpc3NsYXd3dy9lbWJlZC9wb2p2enFaP2hlaWdodD0yNjUmdGhlbWUtaWQ9ZGFyayZkZWZhdWx0LXRhYj1odG1sLHJlc3VsdCIgZnJhbWVib3JkZXI9Im5vIiBhbGxvd3RyYW5zcGFyZW5jeT0idHJ1ZSIgYWxsb3dmdWxsc2NyZWVuPSJ0cnVlIiBsb2FkaW5nPSJsYXp5Ij4KICBTZWUgdGhlIFBlbiA8YSBocmVmPSdodHRwczovL2NvZGVwZW4uaW8vdmxhZGlzc2xhd3d3L3Blbi9wb2p2enFaJz5saWdodCBidWxiPC9hPiBieSB2bGFkaXNzbGF3dwogICg8YSBocmVmPSdodHRwczovL2NvZGVwZW4uaW8vdmxhZGlzc2xhd3d3Jz5AdmxhZGlzc2xhd3d3PC9hPikgb24gPGEgaHJlZj0naHR0cHM6Ly9jb2RlcGVuLmlvJz5Db2RlUGVuPC9hPi4KPC9pZnJhbWU+",
  "page-url": "/story/05066b94-784b-4913-971b-c01a1bcfb43d/element/5bf7ba6d-9a51-41f2-b919-47539c4a5acd",
  type: "jsembed",
  "family-id": "f564d440-c969-4c35-9c5e-ddb032e22b05",
  title: "",
  id: "5bf7ba6d-9a51-41f2-b919-47539c4a5acd",
  metadata: {},
  subtype: null
};

storiesOf("Embed", module)
  .addDecorator((story) => (
    <Layout story={textStory} config={config}>
      {story()}
    </Layout>
  ))
  .add("JsEmbed with iframe", () => <Embed element={sampleJsEmbedElement} />);
