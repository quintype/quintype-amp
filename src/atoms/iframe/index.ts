import { Iframe } from "./iframe";
export { Iframe };
