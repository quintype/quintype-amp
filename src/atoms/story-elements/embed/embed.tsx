import React from "react";
import { StoryElementProps } from "../types";
import { Iframe } from "../../iframe";
import atob from "atob-utf-8";
import { withStoryAndConfig } from "../../../context";
import get from "lodash.get";

export const DefaultEmbed = ({ element }: StoryElementProps) => {
  const embedData = element["embed-js"] ? atob(element["embed-js"]) : "";
  const src = getIframeSourceURL(embedData);
  const title = element.subtype || element.title || "";
  return src ? <Iframe src={encodeURI(src)} title={title} /> : null;
};

export const getIframeSourceURL = (str: string): string | null => {
  if (/iframe/.test(str)) {
    const regExecArray = /(?<=src=").*?(?=[*"])/.exec(str);
    return regExecArray && regExecArray.length > 0 ? regExecArray[0] : null;
  }
  return null;
};

export const EmbedBase = ({ element, story, config }: StoryElementProps) => {
  const embedRender = get(config, ["opts", "render", "storyElementRender", "embedRender"], null);
  return embedRender ? embedRender({ story, config, element }) : <DefaultEmbed element={element} />;
};

/**
 * Embed is a story element.
 * The look of the Embed can be changed using the render prop embedRender. In case embedRender is passed in the config, it is rendered. Otherwise a default Embed is rendered.
 *
 * @param {Object} params object containing parameters passed to the render prop
 * @param {Object} params.story story object
 * @param {Object} params.config config object
 * @param {Object} params.element the story element. This is same as the story element found in the story API response
 *
 * ```javascript
 * ...
 * ampRoutes(app, {
 *  render: {
 *    storyElementRender: {
 *      embedRender: ({ story, config, element }) => <MyCustomEmbed story={story} config={config} storyElement={element} />
 *    }
 *  }
 * })
 * ...
 * ```
 *
 * @category StoryElements
 * @module Embed
 * @component
 */

export const Embed = withStoryAndConfig(EmbedBase);
