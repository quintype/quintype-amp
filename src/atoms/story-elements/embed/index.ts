import { Embed } from "./embed";

export { Embed };
