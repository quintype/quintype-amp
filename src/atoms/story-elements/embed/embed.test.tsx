import React from "react";
import { getIframeSourceURL, EmbedBase, DefaultEmbed } from "./embed";
import { shallow } from "enzyme";
import { Iframe } from "../../iframe";
import { config, textStory } from "../../../__fixtures__";

const sampleJsEmbedElement = {
  description: "",
  "embed-js":
    "PGlmcmFtZSBoZWlnaHQ9IjI2NSIgc3R5bGU9IndpZHRoOiAxMDAlOyIgc2Nyb2xsaW5nPSJubyIgdGl0bGU9ImxpZ2h0IGJ1bGIiIHNyYz0iaHR0cHM6Ly9jb2RlcGVuLmlvL3ZsYWRpc3NsYXd3dy9lbWJlZC9wb2p2enFaP2hlaWdodD0yNjUmdGhlbWUtaWQ9ZGFyayZkZWZhdWx0LXRhYj1odG1sLHJlc3VsdCIgZnJhbWVib3JkZXI9Im5vIiBhbGxvd3RyYW5zcGFyZW5jeT0idHJ1ZSIgYWxsb3dmdWxsc2NyZWVuPSJ0cnVlIiBsb2FkaW5nPSJsYXp5Ij4KICBTZWUgdGhlIFBlbiA8YSBocmVmPSdodHRwczovL2NvZGVwZW4uaW8vdmxhZGlzc2xhd3d3L3Blbi9wb2p2enFaJz5saWdodCBidWxiPC9hPiBieSB2bGFkaXNzbGF3dwogICg8YSBocmVmPSdodHRwczovL2NvZGVwZW4uaW8vdmxhZGlzc2xhd3d3Jz5AdmxhZGlzc2xhd3d3PC9hPikgb24gPGEgaHJlZj0naHR0cHM6Ly9jb2RlcGVuLmlvJz5Db2RlUGVuPC9hPi4KPC9pZnJhbWU+",
  "page-url": "/story/05066b94-784b-4913-971b-c01a1bcfb43d/element/5bf7ba6d-9a51-41f2-b919-47539c4a5acd",
  type: "jsembed",
  "family-id": "f564d440-c969-4c35-9c5e-ddb032e22b05",
  title: "",
  id: "5bf7ba6d-9a51-41f2-b919-47539c4a5acd",
  metadata: {},
  subtype: null
};

const sampleJsEmbedElementWithoutIframe = {
  description: "",
  "embed-js":
    "PGJsb2NrcXVvdGUgY2xhc3M9InR3aXR0ZXItdHdlZXQiPjxwIGxhbmc9ImVuIiBkaXI9Imx0ciI+UmVzaXplT2JzZXJ2ZXIgbGV0cyB1cyBvYnNlcnZlIGNoYW5nZXMgdG8gdGhlIGxheW91dCBzaXplIG9mIGVsZW1lbnRzOiA8YSBocmVmPSJodHRwczovL3QuY28vWklzd1F5TmVtVCI+aHR0cHM6Ly90LmNvL1pJc3dReU5lbVQ8L2E+IEl0JiMzOTtzIGluIHRoZSBsYXRlc3QgU2FmYXJpISA8YnI+PGJyPk5vdyBpbiAqYWxsKiBtb2Rlcm4gYnJvd3NlcnMsIHdlIGNhbiBlZmZpY2llbnRseSBkZXNpZ24gY29tcG9uZW50cyByZXNwb25zaXZlIHRvIHRoZWlyIGNvbnRhaW5lciB2cy4ganVzdCB0aGUgdmlld3BvcnQuIDxhIGhyZWY9Imh0dHBzOi8vdC5jby9UTks5VkRqU0FaIj5waWMudHdpdHRlci5jb20vVE5LOVZEalNBWjwvYT48L3A+Jm1kYXNoOyBBZGR5IE9zbWFuaSAoQGFkZHlvc21hbmkpIDxhIGhyZWY9Imh0dHBzOi8vdHdpdHRlci5jb20vYWRkeW9zbWFuaS9zdGF0dXMvMTI0ODg5OTA4NjQ5MDU1ODQ2NT9yZWZfc3JjPXR3c3JjJTVFdGZ3Ij5BcHJpbCAxMSwgMjAyMDwvYT48L2Jsb2NrcXVvdGU+IDxzY3JpcHQgYXN5bmMgc3JjPSJodHRwczovL3BsYXRmb3JtLnR3aXR0ZXIuY29tL3dpZGdldHMuanMiIGNoYXJzZXQ9InV0Zi04Ij48L3NjcmlwdD4=",
  "page-url": "/story/05066b94-784b-4913-971b-c01a1bcfb43d/element/5bf7ba6d-9a51-41f2-b919-47539c4a5acd",
  type: "jsembed",
  "family-id": "f564d440-c969-4c35-9c5e-ddb032e22b05",
  title: "",
  id: "5bf7ba6d-9a51-41f2-b919-47539c4a5acd",
  metadata: {},
  subtype: null
};

describe("Embed", () => {
  it("should render embed", () => {
    const wrapper = shallow(<DefaultEmbed element={sampleJsEmbedElement} />);
    expect(wrapper.find(Iframe).length).toBe(1);
  });
  it("should render null", () => {
    const wrapper = shallow(<DefaultEmbed element={sampleJsEmbedElementWithoutIframe} />);
    expect(wrapper.find(Iframe).length).toBe(0);
  });
  it("should call storyElementRender prop when passed to opts", () => {
    const storyElementRender = jest.fn();
    const modifiedConfig = { ...config, opts: { ...config.opts, storyElementRender } };
    const wrapper = shallow(<EmbedBase element={sampleJsEmbedElement} story={textStory} config={modifiedConfig} />);
    expect(storyElementRender.mock.calls.length).toBe(1);
    expect(wrapper.find(DefaultEmbed).length).toBe(0);
  });
});

describe("getIframeSourceURL", () => {
  it("should return source of iframe", () => {
    expect(getIframeSourceURL('<iframe frameborder="0" src="https://www.google.com" />')).toBe(
      "https://www.google.com"
    );
  });
  it("should return null if source is missing", () => {
    expect(getIframeSourceURL('<script src="www.google.com" />')).toBe(null);
  });
});
