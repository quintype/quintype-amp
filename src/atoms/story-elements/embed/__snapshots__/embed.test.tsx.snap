// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Embed should render default 1`] = `""`;
